### PR TITLE
Jplalor/eval

### PR DESCRIPTION
**Pull request by jplalor**
- Implement evaluate cli command
- Add predict method to IRT models
- Update datasets to track whether observations should be included in training

### CODE REVIEW

**Comment by jplalor** [comment]
@EntilZha Can you take a look at this PR when you get a chance?

File: py_irt/cli.py [hunk 1]
```diff
@@ -36,9 +40,123 @@ def train(
...
+    if evaluation == "heldout":
+        with open(data_path) as f:
+            items = []
+            for line in f:
+                submission = json.loads(line)
+                model_id = submission["subject_id"]
+                for example_id in submission["responses"].keys():
+                    items.append((model_id, example_id))
+            random.seed(seed)
+            train, test = train_test_split(items, train_size=train_size)
```
Comment (EntilZha) [inline py_irt/cli.py:67]: Should this take in a random seed, or does the call before handle that?

File: py_irt/cli.py [hunk 2]
```diff
@@ -36,9 +40,123 @@ def train(
...
+                training_dict[model_id][example_id] = True 
+            for model_id, example_id in test:
+                training_dict.setdefault(model_id, dict())
+                training_dict[model_id][example_id] = False 
+        dataset = Dataset.from_jsonlines(data_path, train_items=training_dict)
+    else:
+        dataset = Dataset.from_jsonlines(data_path)
+
+    # deep copy for training
+    training_data = copy.deepcopy(dataset)
```
Comment (EntilZha) [inline py_irt/cli.py:80]: Is the dataset mutated in some way that matters?

File: py_irt/cli.py [hunk 3]
```diff
@@ -36,9 +40,123 @@ def train(
...
+    # filter out test data
+    console.log("Evaluating Model...")
+    testing_idx = [i for i in range(len(dataset.training_example)) if not dataset.training_example[i]]
+    if len(testing_idx) > 0:
+        dataset.observation_subjects = [dataset.observation_subjects[i] for i in testing_idx]
+        dataset.observation_items= [dataset.observation_items[i] for i in testing_idx]
+        dataset.observations = [dataset.observations[i] for i in testing_idx]
+        dataset.training_example = [dataset.training_example[i] for i in testing_idx]
+
+    preds = trainer._irt_model.predict(dataset.observation_subjects, dataset.observation_items)
```
Comment (EntilZha) [inline py_irt/cli.py:98]: We should probably make the model `trainer.irt_model` instead of `trainer._irt_model` if we're calling it here

File: py_irt/cli.py [hunk 4]
```diff
@@ -36,9 +40,123 @@ def train(
...
+    print(preds)
```
Comment (EntilZha) [inline py_irt/cli.py:99]: Is the print here necessary?

File: py_irt/cli.py [hunk 5]
```diff
@@ -36,9 +40,123 @@ def train(
...
+    console.log("Evaluating Model...")
+    # load saved params
+    irt_params = read_json(parameter_path)
+
+    # load subject, item pairs we want to test
+    subject_item_pairs = read_jsonlines(test_pairs_path)
+
+    # calculate predictions and write them to disk
+    config = IrtConfig(model_type=model_type, epochs=epochs, initializers=initializers)
+    _irt_model = IRT_MODELS[model_type](
```
Comment (EntilZha) [inline py_irt/cli.py:138]: why not just `irt_model`?

File: py_irt/models/four_param_logistic.py [hunk 6]
```diff
@@ -199,6 +201,18 @@ def export(self):
             "lambdas": pyro.param("lambdas").data.tolist(),
         }
 
+    def predict(self, models, items, params_from_file=None):
```
Comment (EntilZha) [inline py_irt/models/four_param_logistic.py:204]: I think we should be consistent and use `subjects` here

File: py_irt/models/one_param_logistic.py [hunk 7]
```diff
@@ -211,6 +213,16 @@ def fit_MCMC(self, models, items, responses, num_epochs):
         print(theta_sum)
         print(b_sum)
 
+    def predict(self, models, items, params_from_file=None):
```
Comment (EntilZha) [inline py_irt/models/one_param_logistic.py:216]: I think we should be consistent and use `subjects` here

File: py_irt/models/one_param_logistic.py [hunk 8]
```diff
@@ -211,6 +213,16 @@ def fit_MCMC(self, models, items, responses, num_epochs):
...
+        "predict p(correct | params) for a specified list of model, item pairs"
```
Comment (EntilZha) [inline py_irt/models/one_param_logistic.py:217]: For doc comments, this should be triple quote

File: py_irt/models/four_param_logistic.py [hunk 9]
```diff
@@ -199,6 +201,18 @@ def export(self):
...
+        "predict p(correct | params) for a specified list of model, item pairs"
```
Comment (EntilZha) [inline py_irt/models/four_param_logistic.py:205]: Doc comments are triple quote

File: py_irt/models/two_param_logistic.py [hunk 10]
```diff
@@ -271,6 +273,17 @@ def fit_MCMC(self, models, items, responses, num_epochs):
         print(theta_sum)
         print(b_sum)
 
+    def predict(self, models, items, params_from_file=None):
```
Comment (EntilZha) [inline py_irt/models/two_param_logistic.py:276]: Ditto on subjects and doc comment being triple quote

File: py_irt/dataset.py [hunk 11]
```diff
@@ -40,7 +42,7 @@ def get_item_accuracies(self) -> Dict[str, ItemAccuracy]:
         return item_accuracies
 
     @classmethod
-    def from_jsonlines(cls, data_path: Path):
+    def from_jsonlines(cls, data_path: Path, train_items: dict = None):
```
Comment (EntilZha) [inline py_irt/dataset.py:45]: This is fine for now, but we should allow a way to specify this as a data file so we can support train in command, persisting everything to disk, and then eval in another.

**Review by EntilZha — APPROVED** [review]
Made a few comments, LGTM overall and will be helpful to UMD IRT QA team to have an easy way to evaluate models

File: py_irt/cli.py (hunk 1 above)
Comment (jplalor) [reply py_irt/cli.py]: I updated this to pass the seed directly to train_test_split.

File: py_irt/cli.py (hunk 2 above)
Comment (jplalor) [reply py_irt/cli.py:79]: Yes, the validation items are removed (see training.py line 64). There's probably a cleaner way to do this (with a mask or something like it?).

File: py_irt/cli.py (hunk 3 above)
Comment (jplalor) [reply py_irt/cli.py]: Agree, this has been updated throughout.

File: py_irt/cli.py (hunk 4 above)
Comment (jplalor) [reply py_irt/cli.py]: No, removed.

File: py_irt/cli.py (hunk 5 above)
Comment (jplalor) [reply py_irt/cli.py]: Updated (similar to other comment).

File: py_irt/models/four_param_logistic.py (hunk 6 above)
Comment (jplalor) [reply py_irt/models/four_param_logistic.py]: Agree, updated throughout.

File: py_irt/models/one_param_logistic.py (hunk 7 above)
Comment (jplalor) [reply py_irt/models/one_param_logistic.py]: Agree, updated throughout.

File: py_irt/models/one_param_logistic.py (hunk 8 above)
Comment (jplalor) [reply py_irt/models/one_param_logistic.py]: Updated throughout.

File: py_irt/models/four_param_logistic.py (hunk 9 above)
Comment (jplalor) [reply py_irt/models/four_param_logistic.py]: Updated throughout.

File: py_irt/models/two_param_logistic.py (hunk 10 above)
Comment (jplalor) [reply py_irt/models/two_param_logistic.py]: Updated throughout.
